frontend/lib/auditlogquery/index.ts

```diff
@@ -0,0 +1,200 @@
+    } else if (where.maxCreatedAt) {
+      components.push(`MaxCreatedAt:${where.maxCreatedAt}`)
+    } else {
+      console.warn(new Error(`Unknown where: ${JSON.stringify(where)}`))
+    }
+  }
+  return components.join(',')
+}
+
+const decodeAudtLogQueryWheres = (wheres: string): AuditLogQueryWhere[] => {
+  const components = wheres.split(',')
+  const result: AuditLogQueryWhere[] = []
+  for (const component of components) {
+    if (component === '') {
+      continue
+    }
+    const [key, value] = component.split(':')
+    switch (key) {
+      case 'Action':
+        result.push({
+          inAction: value.split('|') as AuditLogQueryWhere['inAction'],
+        })
+        break
+      case 'ActorId':
+        result.push({
+          inActorId: value.split('|') as AuditLogQueryWhere['inActorId'],
+        })
+        break
+      case 'ActorType':
+        result.push({
+          inActorType: value.split('|') as AuditLogQueryWhere['inActorType'],
+        })
+        break
+      case 'ActorOwnerId':
+        result.push({
+          inActorOwnerId: value.split('|') as AuditLogQueryWhere['inActorOwnerId'],
+        })
+        break
+      case 'Id':
+        result.push({
+          inId: value.split('|') as AuditLogQueryWhere['inId'],
+        })
+        break
+      case 'TargetType':
+        result.push({
+          inTargetType: value.split('|') as AuditLogQueryWhere['inTargetType'],
+        })
+        break
+      case 'TargetId':
+        result.push({
+          inTargetId: value.split('|') as AuditLogQueryWhere['inTargetId'],
+        })
+        break
+      case 'TargetOwnerId':
+        result.push({
+          inTargetOwnerId: value.split('|') as AuditLogQueryWhere['inTargetOwnerId'],
+        })
+        break
+      case 'MinCreatedAt':
+        result.push({
+          minCreatedAt: value,
+        })
+        break
+      case 'MaxCreatedAt':
+        result.push({
+          maxCreatedAt: value,
+        })
+        break
+      default:
+        console.warn(new Error(`Unknown where: ${JSON.stringify(key)}`))
+    }
+  }
+  return result
+}
+
+const encodeAuditLogQueryLimit = (limit: number): string => {
+  if (limit === limitDefault) {
+    return ''
+  }
+  return `${limit}`
+}
+
+const decodeAuditLogQueryLimit = (limit: string): number => {
+  if (limit === '') {
+    return limitDefault
+  }
+  return parseInt(limit)
+}
+
+const encodeAuditLogQueryCursor = (cursor: string): string => {
+  return encodeURIComponent(cursor)
+}
+
+const decodeAuditLogQueryCursor = (cursor: string): string => {
+  return decodeURIComponent(cursor)
+}
+
+const sortsQP = 's'
+const wheresQP = 'w'
+const limitQP = 'l'
+const limitDefault = 100
+const cursorQP = 'c'
+const pageURLBase = '/auditlog'
+
+export const urlReactiveAuditLogQuery = (fromQueryReactiveWithDefault: (key: string, defaultValue: string) => WritableComputedRef<string>): WritableComputedRef<AuditLogQueryReq> => {
```

Comment:
You're playing a dangerous game, making library functions that depend on access to things that use `setup`-only constructs like `computed`. It's already two levels of indirection (`urlReactiveAuditLogQuery -> fromQueryReactiveWithDefault -> computed`), it just makes it too easy to forget you should only be calling it in a `setup` function.

It looks like this will only ever be used once, you could just inline it into `/audit-logs` or into a component used by `/audit-logs` where it's unambiguously only ever called from `setup`

And more generally, I think this pattern of taking in functions that should only ever be called in `setup` is dicey, for the above mentioned reason that it obfuscates where things are being called.

Reply:
I think this isn't quite correct: I'm pretty sure (from experience around how we do editors, ex) that `computed` can be created and used outside of `setup`. One of the vue authors [seems to hold this position too](https://github.com/vuejs/core/discussions/9779#discussioncomment-7797404). Could you provide documentation for exactly what can and can't be called outside of setup? A 5 minute google left me without a clear north star.

By taking `fromQueryReactiveWithDefault` as an argument rather than invoking the composable to pull it out, we avoid invoking a composable (which I agree is a riskier game). 

My main goal was just keeping the logic for constructing + decoding URLs that point to audit log queries in one place, to avoid exporting the ~10 helper methods and constants that would be needed to replicate this interface outside of this package. I agree that we shouldn't overly obfuscate where/when core functionality is being used, but I think this approach actually does a good job of balancing between the competing goals as I understand them.